src/Javinizer/Private/Convert-JVTitle.ps1

```diff
@@ -161,34 +167,17 @@ function Convert-JVTitle {
         # Clean any trailing text if not removed by $RemoveStrings
         for ($x = 0; $x -lt $fileBaseNameUpper.Length; $x++) {
             $filePartNumber = $null
-            #Match ID-###A, ID###B, etc.
-            if ($fileBaseNameUpper[$x] -match "[-][0-9]{1,6}Z?E?[a-dA-D]") {
-                $fileP1, $fileP2, $fileP3 = $fileBaseNameUpper[$x] -split "([-][0-9]{1,6}Z?E?)"
-                $fileBaseNameUpperCleaned += $fileP1 + $fileP2
-                if ($fileP3 -eq 'A') { $filePartNumber = '1' }
-                elseif ($fileP3 -eq 'B') { $filePartNumber = '2' }
-                elseif ($fileP3 -eq 'C') { $filePartNumber = '3' }
-                elseif ($fileP3 -eq 'D') { $filePartNumber = '4' }
-                #elseif ($fileP3 -eq 'E') { $filePartNumber = '5' }
-                #elseif ($fileP3 -eq 'F') { $filePartNumber = '6' }
-                #elseif ($fileP3 -eq 'G') { $filePartNumber = '7' }
-                #elseif ($fileP3 -eq 'H') { $filePartNumber = '8' }
-                #elseif ($fileP3 -eq 'I') { $filePartNumber = '9' }
-            }
+            # Match ID-###A, ID###B, etc.
             # Match ID-###-A, ID-###-B, etc.
-            elseif ($fileBaseNameUpper[$x] -match "[-][0-9]{1,6}Z?E?[-][a-dA-D]") {
+            # Match ID-### - A, ID-### - B, etc.
+            if ($fileBaseNameUpper[$x] -match "[-][0-9]{1,6}Z?E?\s?[-]?\s?[A-Z]$") {
                 $fileP1, $fileP2, $fileP3 = $fileBaseNameUpper[$x] -split "([-][0-9]{1,6}Z?E?)"
-                $fileBaseNameUpperCleaned += $fileP1 + $fileP2
-                $fileP3 = $fileP3 -replace '-', ''
-                if ($fileP3 -eq 'A') { $filePartNumber = '1' }
-                elseif ($fileP3 -eq 'B') { $filePartNumber = '2' }
-                elseif ($fileP3 -eq 'C') { $filePartNumber = '3' }
-                elseif ($fileP3 -eq 'D') { $filePartNumber = '4' }
-                #elseif ($fileP3 -eq 'E') { $filePartNumber = '5' }
-                #elseif ($fileP3 -eq 'F') { $filePartNumber = '6' }
-                #elseif ($fileP3 -eq 'G') { $filePartNumber = '7' }
-                #elseif ($fileP3 -eq 'H') { $filePartNumber = '8' }
-                #elseif ($fileP3 -eq 'I') { $filePartNumber = '9' }
+                $fileBaseNameUpperCleaned += $fileP1 + "-" + (($fileP2 -replace '-', '') -replace '^0{1,5}', '').PadLeft(3, '0')
+                $fileP3 = ($fileP3 -replace '-', '').Trim()
+                $asciiP3 = [int][char]$fileP3
+                if ($asciiP3 -gt 64 -and $asciiP3 -lt 91) {
+                    $filePartNumber = $asciiP3 - 64
+                }

```

Comment:
My reasoning for only including letter matches up to D was because of edge cases where the movie ID actually contains a letter.

For example, `IBW-770z / ibw00770z` or `KTRA-231e / ktra00231e`. I know you kept in the logic where if it matches `Z` or `E`, it won't assign a partnumber or remove it, but I think it does cause consistency issues if a user actually has a multi-part movie that goes to E and doesn't get assigned a part number.

```
Id        ContentId  FileName      BaseName  Directory    FullName                   Extension Length PartNumber
--        ---------  --------      --------  ---------    --------                   --------- ------ ----------
ABP-420   ABP00420   ABP-420A.mp4  ABP-420A  C:\Javinizer C:\Javinizer\ABP-420A.mp4  .mp4           0          1
ABP-420   ABP00420   ABP-420B.mp4  ABP-420B  C:\Javinizer C:\Javinizer\ABP-420B.mp4  .mp4           0          2
ABP-420   ABP00420   ABP-420C.mp4  ABP-420C  C:\Javinizer C:\Javinizer\ABP-420C.mp4  .mp4           0          3
ABP-420   ABP00420   ABP-420D.mp4  ABP-420D  C:\Javinizer C:\Javinizer\ABP-420D.mp4  .mp4           0          4
ABP-420E  ABP00420E  ABP-420E.mp4  ABP-420E  C:\Javinizer C:\Javinizer\ABP-420E.mp4  .mp4           0
ABP-420   ABP00420   ABP-420F.mp4  ABP-420F  C:\Javinizer C:\Javinizer\ABP-420F.mp4  .mp4           0          6
ABP-420   ABP00420   ABP-420G.mp4  ABP-420G  C:\Javinizer C:\Javinizer\ABP-420G.mp4  .mp4           0          7
IBW-230Z  IBW00230Z  IBW-230z.mp4  IBW-230z  C:\Javinizer C:\Javinizer\IBW-230z.mp4  .mp4           0
KTRA-213E KTRA00213E KTRA-213e.mp4 KTRA-213e C:\Javinizer C:\Javinizer\KTRA-213e.mp4 .mp4           0
```

Do you know of any other studios that using `e` or `z` in the movie ID? If not, we could probably just hard-code logic to match both IBW-###z and KTRA-###e so that it'll be consistent for multipart movies up to Z.

Reply:
Also, line 177 is throwing an error when it matches an ID that is suffixed with `e` or `z`.

```powershell
InvalidArgument: C:\test\Javinizer\src\Javinizer\Private\Convert-JVTitle.ps1:177
Line |
 177 |                  $asciiP3 = [int][char]$fileP3
     |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Cannot convert value "" to type "System.Char". Error: "String must be exactly one character long."
```

Reply:
I am presently not aware of any other studios using suffixes like the ones you have just mentioned.

I guess there're a couple of ways to go about this, and if the edge cases are as rare as the ones you mentioned above, maybe hardcoding a list of studios that uses such suffix would be the way to go.

Also, I have also personally not seen series spanning more than 3 parts, so perhaps we can limit the number of parts to A-D as well.

In the meantime, I am considering of curating/scraping a list of all studio's s 品番 to see what are the possible edge cases.  If there are really not that many studios using the suffixes approach, then I think hardcoding would be the way to go.

Reply:
> Also, I have also personally not seen series spanning more than 3 parts, so perhaps we can limit the number of parts to A-D as well.

For regular censored JAV, for probably 99% of the time, it's either a single of two-part movie. And in many cases, they would use a different numbering schema like `_1, _2` etc. In recent memory, I believe I've only really seen VR videos go up to something crazy like 5-7 parts.

> In the meantime, I am considering of curating/scraping a list of all studio's s 品番 to see what are the possible edge cases. If there are really not that many studios using the suffixes approach, then I think hardcoding would be the way to go.

That's a good idea. If we can determine with a good enough confidence level that this is the case where no other studios are using letter suffixes let's do this.

Reply:
@jvlflame I have curated a list of all movies with endings with alphabets. I scraped all of these from DMM's website. There's a possibility that it could be an incomplete list. Nevertheless, it should somewhat be helpful to gauge the number of studios with such labels.

Additionally, since it's DMM, it doesn't include uncensored videos.

It seems that even simple alphabets like `a`, `b` aren't spared. You can find the list from the following gist:

https://gist.github.com/seeyabye/b1ea64f59053bb75a446025865a94d75

The following are the list of studios after applying a unique filter:

```python
array(['1idol', '1bksp', '1iesp', '1ie', '1iegr', '1ievd', '1iean',
       '2cyf', '23xy', '23le', '23pst', '23lv', '23wf', '23ymo', '23mc',
       '23lr', '23bes', '23mso', '23bnr', '23mgj', '23ot', '23ter',
       '23qu', '23tks', '23mch', '23hl', '23sdl', '23ste', '23ucc',
       '23kin', '23sc', '23sl', '23msc', '23uct', '23uco', '23gmk',
       '23ucl', '42vri', '44rda', '46rgdr', '47ar', '48shpdv', '50ktdvr',
       'td', '56srv', '56rto', '57jksr', '57mcsr', '57bdsr', '61rmd',
       '65mkdv', '65tpd', '67gesd', '86axdvd', '86aaxdvd', '86aedvd',
       '125ud', '148dgkd', '148syun', '148mixd', '148abcd', '148pnvi',
       '149atm', '149rd', '150eosd', '150hkid', '150dvd', '150ergd',
       '150emuo', '150igmd', '150tmgz', '150uaid', '151re', '165par',
       '165cd', '165mgr', '171lhtd', '171kprd', '171tad', '171mgrd',
       '171axrd', '171zbrd', '171bwrd', '171bnrd', 'h_', '422esp',
       '422dkk', '71gas', '77dpd', '483zsrd', '483dmx', '483trx',
       '483kdx', '303temp', '303see', '303goku', '303eight', '303hgb',
       '303peach', '504ibw', '504mom', '504aki', '504gkb', '1dandy',
       '1sdds', '1rdandy', '1fset', '1rfset', '1rifdva', '189dcm', 'sm',
       '5433bsjdv', 'bomc', 'oycvr', '5561zipf', '5561sidi', 'ktft'],
      dtype=object)
```

and a couple of strange ones that begin with h_
```python
array(['h_1325g', 'h_1489j', 'h_001nsdd', 'h_001arm', 'h_001bjm',
       'h_005rbc', 'h_006bog', 'h_006ght', 'h_100rmow', 'h_047wa',
       'h_081jjor', 'h_081rllyd', 'h_081rjftd', 'h_081rfrss', 'h_093r',
       'h_094ktra', 'h_115gang', 'h_115seed', 'h_115deed', 'h_115yeed',
       'h_157cat', 'h_205ssnd', 'h_236tkdt', 'h_236tkdg', 'h_236tkdw',
       'h_275tdbr', 'h_275tdmj', 'h_308aoz', 'h_479gah', 'h_574iqpa',
       'h_706higr', 'h_706omama', 'h_706gtrp', 'h_706slro', 'h_706fagn',
       'h_706whan', 'h_706jelly', 'h_706faxx', 'h_706pypy', 'h_706ppmn',
       'h_796much', 'h_1000spye', 'h_1001oyaj', 'h_1002jgaho', 'h_1171r',
       'h_1190vrad', 'h_1221higm', 'h_1291fjs', 'h_456mb', 'h_1266jnob'],
      dtype=object)
```

In total, there're approximately 167 studios that ends with alphabets.

Coming up with a elegant regex would not be easy for this. We can include all these 168 studios as an exceptional case and check if the filenames contain these studio names. We can have this list user-defined like the other csv files.

What do you think?

Reply:
Thanks for putting this together.

I'd say I'm against creating another external file to manage these specific cases. It adds an extra file to maintain and we would also have to match each of the studios up with their DVD IDs as well.

Quickly going through the list, it seems like the movies people are most likely downloading are the ones suffixed with `e`, `r`, and `z`. Changing the multi-part alphabet match back to only `D` and then keeping the `Z?E?` and adding `R?` as part of the regex would probably be the cleanest solution while still matching the majority of movies that people are sorting.

Reply:
Thank you for the suggestion. 

I personally think that would not be the cleanest solution. 

Eventually, you will have users complaining that certain suffixes don't work. If we take a look at the ranking in terms of videos ending in their respective alphabets, `E` and `Z` there're probably not that many in contrast to those ending with `D` and `R`. If we now include D in the list, the number of supported multiparts would reduce to only 3... (A, B and C).

```python
[('d', 5391),
 ('r', 5190),
 ('h', 2382),
 ('b', 2064),
 ('u', 1884),
 ('s', 1836),
 ('a', 1764),
 ('i', 1725),
 ('g', 1614),
 ('z', 1509),
 ('e', 1413),
 ('k', 1368),
 ('w', 1254),
 ('o', 1059),
 ('v', 993),
 ('t', 807),
 ('j', 639),
 ('p', 528),
 ('c', 501),
 ('m', 474),
 ('x', 444),
 ('y', 372),
 ('f', 276),
 ('l', 267),
 ('n', 186),
 ('q', 183)]
```

You brought up a good point about matching ID's with DVD. I have no idea how different they are, but a quick search shows similarity in terms of strings between the DVD's product ID and `videoa`'s, so maybe there's still some ways to match them.

Then again, I've only curated the list from `videoa`'s section of DMM. The list could be way different for DVDs. I'll scrape it and analyze that as well.

In the mean time, I'll do another PR with current suggestion implemented instead.

Reply:
Are you referring to a different list with that output of suffix counts?
Doing a quick parse of the [gist](https://gist.github.com/seeyabye/b1ea64f59053bb75a446025865a94d75) gave me these results

```powershell
PS > ((iwr 'https://gist.githubusercontent.com/seeyabye/b1ea64f59053bb75a446025865a94d75/raw/6c319f6f23040f9021ccb6bbbdc3b2a77c97cd35/productid.txt').rawcontent) -split '\n' | % { ($_)[-1] } | Group-Object | Sort-Object Count -Descending

Count Name                      Group
----- ----                      -----
 1618 r                         {r, r, r, r…}
  497 z                         {z, z, z, z…}
  294 d                         {d, d, d, d…}
  258 s                         {s, s, s, s…}
  159 b                         {b, b, b, b…}
   99 a                         {a, a, a, a…}
   48 c                         {c, c, c, c…}
   41 e                         {e, e, e, e…}
   31 f                         {f, f, f, f…}
   27 …                         {…
   10 k                         {k, k, k, k…}
    2 g                         {g, g}
    1 h                         {h}
    1 l                         {l}
    1 w                         {w}
    1 x                         {x}
```

I'm speaking from personal experience, but I don't think I've downloaded a single video that uses a suffix other than `D`, `E`, or `Z` out of around ~8000 videos. That's why I'm kind of iffy on including any of the other edge cases.

Reply:
My apologies, I made a mistake in the aggregation code. You're correct, `R` and `Z` has the highest frequency count.